primitives/npos-elections/solution-type/src/lib.rs

```diff
@@ -0,0 +1,354 @@
+			if segment.ident == Ident::new("pages", Span::call_site()) {
+				let tokens = &attr.tokens;
+				if let Ok(parsed) = syn::parse2::<syn::ExprParen>(tokens.clone()) {
+					if let Ok(parsed_pages) = parse_parenthesized_number::<u8>(parsed) {
+						pages = Some(parsed_pages);
+					} else {
+						return Err(syn_err("Failed to parse a number into u8 in `#[pages(_)]`"))
+					}
+				} else {
+					return Err(syn_err("Failed to parse a number into u8 in `#[pages(_)]`"))
+				}
+			}
+		}
+	}
+
+	Ok((has_compact, pages))
+}
+
+impl Parse for SolutionDef {
+	fn parse(input: ParseStream) -> syn::Result<Self> {
+		// optional #[compact] and #[pages(_)]
+		let (compact_encoding, voter_pages) = check_attributes(input)?;
+
+		// <vis> struct <name>
+		let vis: syn::Visibility = input.parse()?;
+		let _ = <syn::Token![struct]>::parse(input)?;
+		let ident: syn::Ident = input.parse()?;
+
+		// ::<V, T, W>
+		let _ = <syn::Token![::]>::parse(input)?;
+		let generics: syn::AngleBracketedGenericArguments = input.parse()?;
+
+		if generics.args.len() != 3 {
+			return Err(syn_err("Must provide 3 generic args."))
+		}
+
+		let expected_types = ["VoterIndex", "TargetIndex", "Accuracy"];
+
+		let mut types: Vec<syn::Type> = generics
+			.args
+			.iter()
+			.zip(expected_types.iter())
+			.map(|(t, expected)| match t {
+				syn::GenericArgument::Type(ty) => {
+					// this is now an error
+					Err(syn::Error::new_spanned(
+						ty,
+						format!("Expected binding: `{} = ...`", expected),
+					))
+				},
+				syn::GenericArgument::Binding(syn::Binding { ident, ty, .. }) => {
+					// check that we have the right keyword for this position in the argument list
+					if ident == expected {
+						Ok(ty.clone())
+					} else {
+						Err(syn::Error::new_spanned(ident, format!("Expected `{}`", expected)))
+					}
+				},
+				_ => Err(syn_err("Wrong type of generic provided. Must be a `type`.")),
+			})
+			.collect::<Result<_>>()?;
+
+		let weight_type = types.pop().expect("Vector of length 3 can be popped; qed");
+		let target_type = types.pop().expect("Vector of length 2 can be popped; qed");
+		let voter_type = types.pop().expect("Vector of length 1 can be popped; qed");
+
+		// (<count>)
+		let count_expr: syn::ExprParen = input.parse()?;
+		let count = parse_parenthesized_number::<usize>(count_expr)?;
+
+		Ok(Self {
+			vis,
+			ident,
+			voter_pages,
+			voter_type,
+			target_type,
+			weight_type,
+			count,
+			compact_encoding,
+		})
+	}
+}
+
+fn parse_parenthesized_number<N: std::str::FromStr>(input_expr: syn::ExprParen) -> syn::Result<N>
+where
+	<N as std::str::FromStr>::Err: std::fmt::Display,
+{
+	let expr = input_expr.expr;
+	let expr_lit = match *expr {
+		syn::Expr::Lit(count_lit) => count_lit.lit,
+		_ => return Err(syn_err("Count must be literal.")),
+	};
+	let int_lit = match expr_lit {
+		syn::Lit::Int(int_lit) => int_lit,
+		_ => return Err(syn_err("Count must be int literal.")),
```

Comment:
same span can be improved here

Reply:
can you elaborate on this?

Reply:
currently the error span is call_site so it will not precisely point to the int literal, I think if we use the int literal span the error message will directly point to the int literal so it is more understandable for user.